qubes/storage/zfs.py

```diff
@@ -0,0 +1,2878 @@
+        self,
+        src_volume: qubes.storage.Volume,
+    ) -> "ZFSVolume":
+        """
+        Import a volume from another one.
+
+        Calls `_wipe_and_clone_from()`.
+        """
+        if not self.save_on_stop:
+            # Nothing to import.  This volume will be blown up
+            # next time its owning VM starts up.
+            return self
+        self.log.debug(
+            "Importing volume %s from source %s",
+            self.volume,
+            src_volume,
+        )
+        if not src_volume.save_on_stop:
+            raise NotImplementedError(
+                f"Cannot import from {self.vid} — volumes where"
+                " save_on_stop=False cannot be exported for import"
+            )
+
+        if self.is_dirty():
+            raise qubes.storage.StoragePoolException(
+                f"Cannot import to dirty volume {self.volume} —"
+                " start and stop its owning qube to clean it up"
+            )
+        self.abort_if_import_in_progress()
+        await self._wipe_and_clone_from(src_volume)
+        # Make clean snapshot so this volume can be cloned later.
+        await self._create_revision("after-import-volume")
+        await self._mark_clean()
+        return self
+
+    def abort_if_import_in_progress(self) -> None:
+        """Abort if an import is in progress."""
+        if self.pool.accessor.volume_exists(
+            self.importing_volume_name,
+            log=self.log,
+        ):
+            raise qubes.storage.StoragePoolException(
+                "Import operation in progress on {}".format(self.volume)
+            )
+
+    @qubes.storage.Volume.locked  # type: ignore
+    async def import_data(self, size: int) -> str:
+        """
+        Return a path name that can be `open()`ed.
+
+        Callers generally write data to the object.
+        """
+        self.log.debug("Importing data of size %s into %s", size, self.volume)
+        if self.is_dirty():
+            raise qubes.storage.StoragePoolException(
+                "Cannot import data to dirty volume {} -"
+                " stop the qube using it first".format(self.volume)
+            )
+        self.abort_if_import_in_progress()
+        imp = self.importing_volume_name
+        self.log.debug("Creating volume for import %s with size %s", imp, size)
+        await self.pool.accessor.create_volume_async(
+            imp,
+            size,
+            self._auto_snapshot_policy,
+            log=self.log,
+        )
+        return os.path.join(ZVOL_DIR, imp)
+
+    @qubes.storage.Volume.locked  # type: ignore
+    async def import_data_end(self, success: bool) -> None:
+        """Either commit imported data, or discard temporary volume"""
+        self.log.debug("End of importing data into %s", self.volume)
+        if success:
+            newvol = self.importing_volume_name
+            self.log.debug("Adopting %s from %s", self.volume, newvol)
+            await self._remove_volume_if_exists()
```

Comment:
From a cursory look it doesn't look like I can't do that with `zfs rename`.

I have a choice here:

* (what we used to do) import inplace, destroying the existing data but preserving existing snapshots
* (what we do now per Demi Marie's request) import in a temporary dataset, and move into final place at the end

To rename a volume to its final name, you *have* to destroy the existing one.  No two ways about it.

**There is a janky workaround** which involves taking a snapshot pre-import, importing in-place, and then if success deleting the snapshot, but if not success reverting to the snapshot pre-import.  Would you like me to explore that scenario?

Reply:
Nope.  A clone is a dataset that's been copied (CoW) from a snapshot.  The clone operation does not bring the snapshots of the original one into the cloned dataset.

Should I explore the scenario offered above?  Want to know before I do the squash pre-merge.

Reply:
> Should I explore the scenario offered above?

IIUC there would be a risk that interrupted import (like, qubesd restart, or power failure) could be considered as successful one (for the purpose of starting this very VM), right? You'd need to be careful with marking that import is in progress and revert it appropriately.
For cloning from other ZFS volumes, that isn't a problem (currently, this is used only on qvm-clone, and the target is a fresh VM anyway).  But if making data import revertable is easy to do, it may be useful. If that would complicate stuff too much, or would be risky to do, then it isn't worth it (personally I probably used this feature once, and I had an alternative anyway).

Reply:
I think it would complicate things too much.  Will ponder over the next months if it's worth doing.

Reply:
It is worth doing, IMO.  Otherwise trying to start the incomplete VM will likely result in extremely confusing errors.

Reply:
That isn't the risk with the current implementation, but _potentially_ (in case of a code bug) with the proposed alternative. The current approach only has a  downside of discarding previous snapshots in case of _successful_ import (so - successful import cannot be reverted).

Reply:
Okay, that makes sense.